spices/SPICE-0001-import-graph-analyzer-api.adoc

```diff
@@ -0,0 +1,336 @@
+=== Standard library module `pkl:analyze`
+
+A new standard library module is introduced.
+It contains the class `ImportAnalyzer`, which provides import graph analysis.
+
+[source,pkl]
+----
+/// Utilities for static analysis of Pkl modules.
+///
+/// Unlike `pkl:reflect`, these methods do not require loading the modules for analysis.
+@ModuleInfo { minPklVersion = "0.26.0" }
+@Since { version = "0.26.0" }
+module pkl.analyze
+
+class ImportAnalyzer {
+  /// The modules to analyze for imports.
+  moduleUris: Listing<String>
+
+  /// The graph of imports declared by [moduleUris].
+  /// Each key-value pair is a module, corresponding with the set of modules
+  /// imported by that module.
+  ///
+  /// The set of all modules in the graph can be obtained via its keys.
+  external imports: Mapping<Uri, Listing<Uri>(isDistinct)>
+
+  /// Mappings of modules from their in-language URI, to the URI that is used to fetch the module's contents.
+  external resolvedUris: Mapping<Uri, Uri>(keys == imports.keys)
+}
+----
+
+The `ImportAnalyzer` is the in-language representation of the `pkl analyze imports` CLI command.
+
+=== Java API
+
+A Java-only API is introduced to perform actual analysis.
+
+Unlike the CLI, this API only parses a module, and does not involve the evaluator.
+
+[source,java]
+----
+public interface ImportAnalyzer {
+    Results analyze(List<ModuleSource> moduleSources);
+
+    interface Results {
+        Map<URI, Set<URI>> getImports();
+
+        Map<URI, URI> getResolvedUris();
+    }
+}
+
+----
+
+=== Example
+
+This Pkl file:
+
+./path/to/my/foo.pkl
+[source,pkl]
+----
+import "@bar/Bar.pkl" // <1>
+import "package://example.com/bar@1.0.0#/Bar.pkl" // <2>
+import "pkl:reflect" // <3>
+import "..." // <4>
+import* "subdir/*.pkl" <5>
+
+res = if (foo) import("foo1.pkl") else import("foo2.pkl") // <6>
+----
+<1> local dependency import of package://example.com/bar@1.0.0
+<2> direct package import
+<3> stdlib import
+<4> triple-dot import (resolves to /path/to/foo.pkl)
+<5> globbed import (resolves to /path/to/my/subdir/foo.pkl)
+<6> conditional import; both are added to the dependency graph
+
+Produces the following result:
+
+[source,pkl]
+----
+imports {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] {} // <1>
+  ["file:///foo/bar/Bar.pkl"] {}
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] {}
+  ["pkl:reflect"] {
+    "pkl:base"
+  }
+  ["pkl:base"] {
+    "pkl:jsonnet"
+    "pkl:xml"
+    "pkl:protobuf"
+  }
+  ["pkl:jsonnet"] {}
+  ["pkl:xml"] {}
+  ["pkl:protobuf"] {
+    "pkl:reflect"
+  }
+  ["file:///path/to/foo.pkl"] {}
+  ["file:///path/to/my/foo1.pkl"] {}
+  ["file:///path/to/my/foo2.pkl"] {}
+  ["file:///path/to/my/subdir/foo.pkl"] {}
+}
+resolvedUris {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] = "file:///path/to/local/bar/Bar.pkl" // <2>
+  ["file:///foo/bar/Bar.pkl"] = "file:///foo/bar/bar.pkl" // <3>
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] = "package://example.com/bar@1.0.0#/Bar.pkl" // <4>
```

Comment:
I thought about that, but I can't find a URI that represents "a file within a zip file downloaded from HTTPS".

Java has jar-URIs, e.g. `jar:https://something.com/foo.zip!myFile.pkl`, but this is a Java-ism and not something generally used by other tools. Also, these URIs start with "jar" and not "zip", which is especially tied to Java.

Reply:
Yeah, you are right that it would be misleading to generate a URI that wouldn't point to the file in question. Leaving it as `package` would signal that you need to understand this scheme to properly fetch the module.